src/middleware/authorisation.coffee

```diff
@@ -75,10 +75,24 @@ exports.authorise = (ctx, done) ->
       pat = new RegExp channel.urlPattern
       # if url pattern matches
       if pat.test ctx.request.url
-        matchedRoles = channel.allow.filter (element) ->
-          return (ctx.authenticated.roles.indexOf element) isnt -1
+        matchedRoles = {}
+        allowedClient = false
+        if ctx.authenticated?
+          # used by messageStore
+          ctx.authenticated.ip = ctx.ip
+
+          if ctx.authenticated.roles?
+            matchedRoles = channel.allow.filter (element) ->
+              return (ctx.authenticated.roles.indexOf element) isnt -1
+          if ((channel.allow.indexOf ctx.authenticated.clientID) isnt -1)
+            allowedClient = true
+        else
+          # used by messageStore
+          ctx.authenticated =
+            ip: ctx.ip
+
         # if the user has a role that is allowed or their username is allowed specifically
-        if matchedRoles.length > 0 or (channel.allow.indexOf ctx.authenticated.clientID) isnt -1
+        if (matchedRoles.length > 0) or allowedClient or ((channel.whitelist.indexOf ctx.ip) isnt -1) or ((channel.authType == 'public') is true)
```

Comment:
I'm not too sure about the whitelist check here. Correct me if I'm wrong, but this means if the client ip is in the whitelist, they will _always_ be authenticated regardless of basic auth or tls certs?

Following our design call around this, my impression was instead that the whitelist is an _additional_ check, so if for example the channel is private, then the client needs to be authenticated via say basic auth _and_ their IP should be in the whitelist.

Maybe we need a check like? (code not tested!)
```
requireWhitelistCheck = channel.whitelist.length > 0
isWhiteListed = not requireWhitelistCheck or ((channel.whitelist.indexOf ctx.ip) isnt -1)
if isWhiteListed and ((matchedRoles.length > 0) or allowedClient or ((channel.authType == 'public') is true))
  ...
```